proof/dleq.go

```diff
@@ -0,0 +1,108 @@
+package proof
+
+import (
+	"errors"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/hash"
+	"github.com/dedis/crypto/random"
+)
+
+// This package provides functionality to create and verify non-interactive
+// zero-knowledge (NIZK) proofs for the equality (EQ) of discrete logarithms (DL).
+// This means, for two values xG and xH one can check that log_{G}(xG) == log_{H}(xH)
+// without revealing the secret value x.
+
+var errorDifferentLengths = errors.New("inputs of different lengths")
+
+// DLEQProof represents a NIZK dlog-equality proof.
+type DLEQProof struct {
+	C  abstract.Scalar // challenge
+	R  abstract.Scalar // response
+	VG abstract.Point  // public commitment with respect to base point G
+	VH abstract.Point  // public commitment with respect to base point H
+}
+
+// NewDLEQProof computes a new NIZK dlog-equality proof by randomly selecting a
+// commitment v, determining the challenge c = H(xG,xH,vG,vH) and the response r
+// = v - cx. It also returns the encrypted base points xG and xH.
+func NewDLEQProof(suite abstract.Suite, G abstract.Point, H abstract.Point, x abstract.Scalar) (proof *DLEQProof, xG abstract.Point, xH abstract.Point, err error) {
+	// Encrypt base points with secret
+	xG = suite.Point().Mul(G, x)
+	xH = suite.Point().Mul(H, x)
+
+	// Commitment
+	v := suite.Scalar().Pick(random.Stream)
+	vG := suite.Point().Mul(G, v)
+	vH := suite.Point().Mul(H, v)
+
+	// Challenge
+	cb, err := hash.Structures(suite.Hash(), xG, xH, vG, vH)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	c := suite.Scalar().Pick(suite.Cipher(cb))
+
+	// Response
+	r := suite.Scalar()
+	r.Mul(x, c).Sub(v, r)
+
+	return &DLEQProof{c, r, vG, vH}, xG, xH, nil
+}
+
+// NewDLEQProofBatch computes lists of NIZK dlog-equality proofs and of
+// encrypted base points xG and xH. Note that the challenge is computed over all
+// input values.
+func NewDLEQProofBatch(suite abstract.Suite, G []abstract.Point, H []abstract.Point, secrets []abstract.Scalar) (proof []*DLEQProof, xG []abstract.Point, xH []abstract.Point, err error) {
+	if len(G) != len(H) || len(H) != len(secrets) {
+		return nil, nil, nil, errorDifferentLengths
```

Comment:
lol that's a very low level nitpick question... I'm not sure, I'm not a fan of named arguments, but here it's needed since we must differ between xG and xH. Since I'm not used to it, I don't know the best practices for that. If you think this is better, why not. In my part, due to my incomplete knowledge, I'd go for the shorter version (so what @Daeinar proposed).